### PR TITLE
Workspace touched date

### PR DESCRIPTION
**Pull request by elioschmutz**
This PR adds the touched date index for workspaces.

The `touched` field was only implemented for dossiers. In this PR, we refactor out the field into its own schema behavior called `ITouched`. We then generally use this behavior for dossiers and workspaces.

The indexer method is now also registered for all `IDexterityContent` objects and will only return the date if the `ITouched` behavior is provided. This makes the implementation more generic and reusable for different content types.

The upgradestep in this PR is not implemented as a nightly upgrade step because migrating the dossiers `touched` filed does not require a reindex, so this is blazing fast. For setting the initial touched date for workspaces we only need to reindex workspaces. There are no deployments with a huge amount of workspaces.

For [CA-4875]

## Checklist

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

_Only applicable should be left and checked._

- Upgrade-Steps:
  - [ ] SQL Operations do not use imported model (see [docs](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/994344994/Upgrade-Steps))
  - [ ] Make it deferrable if possible
  - [ ] Execute as much as possible conditionally
  - DB-Schema migration
    - [ ] All changes on a model (columns, etc) are included in a DB-schema migration.
    - [ ] Constraint names are shorter than 30 characters (`Oracle`)
- API change:
  - [ ] Documentation is updated
  - [ ] API Changelog entry (see [guide](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/451248812/API+Changelog+Guidelines))
  - If breaking:
    - [ ] api-change label added
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- Bug fixed:
  - [ ] Resolved any Sentry issues caused by this bug
- New functionality:
  - [ ] for `document` also works for `mail`
  - [ ] for `task` also works for `forwarding`
- Further improvements needed:
  - [ ] Create follow-up stories and link them in the PR and Jira issue
- [ ] Change could impact client installations, client policies need to be adapted
- New translations
  - [ ] All msg-strings are unicode
- Change in schema definition:
  - [ ] If `missing_value` is specified, then `default` has to be set to the same value

[CA-4875]: https://4teamwork.atlassian.net/browse/CA-4875?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: opengever/base/indexes.py [hunk 1]
```diff
@@ -160,3 +161,8 @@ def getObjPositionInParent(obj):
             return
         return ploneGetObjPositionInParent(obj)()
     return None
+
+
+@indexer(IDexterityContent)
```
Comment (njohner) [inline opengever/base/indexes.py:166]: Why don't you register this for `ITouched`?

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 2]
```diff
@@ -14,6 +14,9 @@ def __call__(self):
     def migrate_touched_to_new_behavior(self):
         query = {'object_provides': 'opengever.dossier.behaviors.dossier.IDossierMarker'}
         for obj in self.objects(query, "Migrate touched from IDossier to ITouched."):
+            if ITouched(obj).touched:
```
Comment (njohner) [inline opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:18]: ? why is that necessary? That won't happen.

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 3]
```diff
@@ -23,3 +24,22 @@ def migrate_touched_to_new_behavior(self):
                 # Reassigning the value to the new behavior is enought and
                 # fast. So there is also no need for a nightly upgrade step.
                 ITouched(obj).touched = dossier_schema.touched
+
+    def index_touched_on_worskspaces(self):
+        query = {'object_provides': 'opengever.workspace.interfaces.IWorkspace'}
+        for obj in self.objects(query, "Index touched on workspaces."):
+            if ITouched(obj).touched:
+                continue  # already migrated
```
Comment (njohner) [inline opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:32]: same here, the upgrade step will not be applied twice, so I don't understand that statement.

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 4]
```diff
@@ -23,3 +24,22 @@ def migrate_touched_to_new_behavior(self):
...
+
+            # Get the last modified object within the workspace or the workspace
+            # itself to determine the current touched date.
+            ITouched(obj).touched = self.catalog_unrestricted_search(
+                {
+                    'path': {'query': '/'.join(obj.getPhysicalPath())},
+                    'sort_on': 'modified',
```
Comment (njohner) [inline opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:39]: I think you should be using `changed` not `modified` here. `modified` is also set when an object is reindexed for example, so it does not contain the information you want IMO.

**Review by njohner — COMMENTED** [review]
Just a few comments, but otherwise LGTM!
Just wondering whether we should also introduce `touched` for `WorkspaceFolder`s?

**Comment by elioschmutz** [comment]
> Just a few comments, but otherwise LGTM! Just wondering whether we should also introduce `touched` for `WorkspaceFolder`s?

Providing it for folders makes no sense at the moment since we do not show the modified/touched date of folders somewhere... (or I missed a view)

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    def migrate_touched_to_new_behavior(self):
+        query = {'object_provides': 'opengever.dossier.behaviors.dossier.IDossierMarker'}
+        for obj in self.objects(query, "Migrate touched from IDossier to ITouched."):
+            dossier_schema = IDossier(obj)
+            if hasattr(dossier_schema, 'touched'):
+                # We do not need to reindex because the field value does not change.
+                # Reassigning the value to the new behavior is enought and
+                # fast. So there is also no need for a nightly upgrade step.
+                ITouched(obj).touched = dossier_schema.touched
```
[the remark this answers is not in the document]
Comment (elioschmutz) [reply opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:26]: You're right 👍  I updated the upgradestep. I now also remove the old value.

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py (hunk 2 above)
Comment (elioschmutz) [reply opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:18]: I removed it

File: opengever/base/indexes.py (hunk 1 above)
Comment (elioschmutz) [reply opengever/base/indexes.py]: 👍 I updated

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            # Get the last modified object within the workspace or the workspace
+            # itself to determine the current touched date.
+            ITouched(obj).touched = self.catalog_unrestricted_search(
+                {
+                    'path': {'query': '/'.join(obj.getPhysicalPath())},
+                    'sort_on': 'changed',
+                    'sort_order': 'descending',
+                    'sort_limit': 1
+                }
+            )[0].modified
```
Comment (njohner) [inline opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:39]: `.changed`, not `.modified`

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    def index_touched_on_worskspaces(self):
+        query = {'object_provides': 'opengever.workspace.interfaces.IWorkspace'}
+        for obj in self.objects(query, "Index touched on workspaces."):
+            # Get the last modified object within the workspace or the workspace
+            # itself to determine the current touched date.
+            ITouched(obj).touched = self.catalog_unrestricted_search(
+                {
+                    'path': {'query': '/'.join(obj.getPhysicalPath())},
+                    'sort_on': 'changed',
```
Comment (njohner) [inline opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:35]: Sorry I just had a closer look, and ToDos don't seem to have a `changed` date. I don't know why though, everything seems to have a changed date except ToDos and ToDoLists (even `meetingagendaitem` has a changed date). So I'm a bit unsure what you should do to set the initial `touched` date...
I'm pretty sure it simply got forgotten to add the Changed date do ToDos... we can have a quick call or discuss that quickly at daily tomorrow.

**Review by njohner — COMMENTED** [review]
I'm still unsure as to the initial value of `touched`. Otherwise LGTM!

File: opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py (hunk 7 above)
Comment (elioschmutz) [reply opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:35]: As discussed: The `IChanged` behavior is missing for todos and todolists. I'll add it and reindex it

Comment (elioschmutz) [reply opengever/core/upgrades/20230110082930_add_i_touched_behavior/upgrade.py:37]: @njohner 👍 updated

**Review by njohner — APPROVED** [review]
Perfect thanks, LGTM
